File: backend/app/schemas/coach.py
"""
Schémas Pydantic v2 — Coach IA SOMA (LOT 9).

Valident les requêtes et sérialisent les réponses des endpoints coach.
"""
from datetime import datetime
from typing import Optional
import uuid

from pydantic import BaseModel, Field


# ── Requêtes ─────────────────────────────────────────────────────────────────

class AskCoachRequest(BaseModel):
    question: str = Field(
        ...,
        min_length=3,
        max_length=2000,
        description="Question posée au coach SOMA.",
        examples=["Pourquoi suis-je aussi fatigué aujourd'hui ?"],
    )
    thread_id: Optional[uuid.UUID] = Field(
        None,
        description="ID du fil de conversation existant (None = nouveau thread).",
    )


class CreateThreadRequest(BaseModel):
    title: Optional[str] = Field(
        None,
        max_length=200,
        description="Titre optionnel du fil de conversation.",
    )


# ── Réponses ─────────────────────────────────────────────────────────────────

class CoachAnswerResponse(BaseModel):
    """Réponse structurée du coach IA."""
    summary: str = Field(..., description="Synthèse courte (1-2 phrases).")
    full_response: str = Field(..., description="Réponse complète en Markdown.")
    recommendations: list[str] = Field(
        default_factory=list,
        description="Liste de recommandations concrètes.",
    )
    warnings: list[str] = Field(
        default_factory=list,
        description="Alertes ou points de vigilance.",
    )
    confidence: float = Field(
        ...,
        ge=0.0,
        le=1.0,
        description="Score de confiance (0-1) basé sur la disponibilité des données.",
    )
    context_tokens_estimate: int = Field(0, description="Estimation du nb de tokens de contexte.")
    model_used: str = Field(..., description="Modèle IA utilisé ou 'mock'.")
    thread_id: uuid.UUID = Field(..., description="ID du fil de conversation.")
    message_id: uuid.UUID = Field(..., description="ID du message coach créé.")


class ConversationThreadResponse(BaseModel):
    """Représentation d'un fil de conversation."""
    id: uuid.UUID
    title: Optional[str]
    summary: Optional[str]
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class ConversationMessageResponse(BaseModel):
    """Représentation d'un message dans une conversation."""
    id: uuid.UUID
    thread_id: uuid.UUID
    role: str  # "user" | "coach"
    content: str
    created_at: datetime
    metadata: Optional[dict] = Field(None, alias="metadata_")

    model_config = {"from_attributes": True, "populate_by_name": True}


class ConversationThreadDetailResponse(BaseModel):
    """Fil de conversation avec ses messages."""
    thread: ConversationThreadResponse
    messages: list[ConversationMessageResponse]


class ThreadListResponse(BaseModel):
    """Liste des fils de conversation."""
    threads: list[ConversationThreadResponse]
    total: int


# ── Quick Advice (LOT 18) — réponse rapide sans persistence ──────────────────

class QuickAdviceRequest(BaseModel):
    """Requête conseil rapide — pas de thread, pas de persistence."""
    question: str = Field(
        ...,
        min_length=3,
        max_length=500,
        description="Question courte posée au coach SOMA.",
        examples=["Dois-je m'entraîner aujourd'hui ?"],
    )


class QuickAdviceResponse(BaseModel):
    """Réponse synthétique du coach — 3 phrases max, sans persistence DB."""
    answer: str = Field(..., description="Réponse principale (2 phrases max).")
    recommendations: list[str] = Field(
        default_factory=list,
        max_length=2,
        description="2 actions concrètes max.",
    )
    alert: Optional[str] = Field(
        None,
        description="1 alerte critique si nécessaire, sinon None.",
    )
    confidence: float = Field(
        ...,
        ge=0.0,
        le=1.0,
        description="Score de confiance basé sur la disponibilité des données.",
    )
    model_used: str = Field(..., description="Modèle IA utilisé ou 'mock'.")
    context_summary: str = Field(
        ...,
        description="Résumé compact du contexte utilisé (ex: readiness: 72%, fatigue: 45%).",
    )

File: mobile/lib/core/models/coach.dart
/// Modèles de données Coach IA SOMA — LOT 9.
///
/// CoachAnswer     : réponse structurée du coach
/// CoachThread     : fil de conversation
/// CoachMessage    : message individuel (user | coach)
library;

// ── Réponse coach ─────────────────────────────────────────────────────────────

class CoachAnswer {
  final String summary;
  final String fullResponse;
  final List<String> recommendations;
  final List<String> warnings;
  final double confidence;
  final int contextTokensEstimate;
  final String modelUsed;
  final String threadId;
  final String messageId;

  const CoachAnswer({
    required this.summary,
    required this.fullResponse,
    required this.recommendations,
    required this.warnings,
    required this.confidence,
    required this.contextTokensEstimate,
    required this.modelUsed,
    required this.threadId,
    required this.messageId,
  });

  factory CoachAnswer.fromJson(Map<String, dynamic> json) {
    return CoachAnswer(
      summary: json['summary'] as String? ?? '',
      fullResponse: json['full_response'] as String? ?? '',
      recommendations: (json['recommendations'] as List<dynamic>?)
              ?.map((e) => e as String)
              .toList() ??
          [],
      warnings: (json['warnings'] as List<dynamic>?)
              ?.map((e) => e as String)
              .toList() ??
          [],
      confidence: (json['confidence'] as num?)?.toDouble() ?? 0.75,
      contextTokensEstimate: json['context_tokens_estimate'] as int? ?? 0,
      modelUsed: json['model_used'] as String? ?? 'mock',
      threadId: json['thread_id'] as String? ?? '',
      messageId: json['message_id'] as String? ?? '',
    );
  }
}

// ── Fil de conversation ───────────────────────────────────────────────────────

class CoachThread {
  final String id;
  final String? title;
  final String? summary;
  final DateTime createdAt;
  final DateTime updatedAt;

  const CoachThread({
    required this.id,
    this.title,
    this.summary,
    required this.createdAt,
    required this.updatedAt,
  });

  factory CoachThread.fromJson(Map<String, dynamic> json) {
    return CoachThread(
      id: json['id'] as String? ?? '',
      title: json['title'] as String?,
      summary: json['summary'] as String?,
      createdAt: json['created_at'] != null
          ? DateTime.parse(json['created_at'] as String)
          : DateTime.now(),
      updatedAt: json['updated_at'] != null
          ? DateTime.parse(json['updated_at'] as String)
          : DateTime.now(),
    );
  }

  /// Label d'affichage : titre ou date formatée.
  String get displayTitle {
    if (title != null && title!.isNotEmpty) return title!;
    final d = createdAt;
    return '${d.day.toString().padLeft(2, '0')}/${d.month.toString().padLeft(2, '0')}/${d.year}';
  }
}

// ── Message de conversation ───────────────────────────────────────────────────

class CoachMessage {
  final String id;
  final String threadId;
  final String role; // 'user' | 'coach'
  final String content;
  final DateTime createdAt;
  final Map<String, dynamic>? metadata;

  const CoachMessage({
    required this.id,
    required this.threadId,
    required this.role,
    required this.content,
    required this.createdAt,
    this.metadata,
  });

  bool get isUser => role == 'user';
  bool get isCoach => role == 'coach';

  factory CoachMessage.fromJson(Map<String, dynamic> json) {
    return CoachMessage(
      id: json['id'] as String? ?? '',
      threadId: json['thread_id'] as String? ?? '',
      role: json['role'] as String? ?? 'user',
      content: json['content'] as String? ?? '',
      createdAt: json['created_at'] != null
          ? DateTime.parse(json['created_at'] as String)
          : DateTime.now(),
      metadata: json['metadata'] as Map<String, dynamic>?,
    );
  }
}

// ── Quick prompts ─────────────────────────────────────────────────────────────

class CoachQuickPrompt {
  final String label;
  final String prompt;
  final String icon;

  const CoachQuickPrompt({
    required this.label,
    required this.prompt,
    required this.icon,
  });
}

const kCoachQuickPrompts = [
  CoachQuickPrompt(
    label: 'Analyse ma journée',
    prompt: 'Analyse ma journée d\'aujourd\'hui et dis-moi ce que je dois améliorer.',
    icon: '📊',
  ),
  CoachQuickPrompt(
    label: 'Dois-je m\'entraîner ?',
    prompt: 'Est-ce que je dois m\'entraîner aujourd\'hui compte tenu de mon état de récupération ?',
    icon: '💪',
  ),
  CoachQuickPrompt(
    label: 'Que manger ce soir ?',
    prompt: 'Que dois-je manger ce soir pour atteindre mes objectifs nutritionnels du jour ?',
    icon: '🥗',
  ),
  CoachQuickPrompt(
    label: 'Pourquoi suis-je fatigué ?',
    prompt: 'Pourquoi suis-je aussi fatigué aujourd\'hui ? Analyse mes données.',
    icon: '😴',
  ),
];
